signer/src/schnorr.rs

```diff
@@ -0,0 +1,156 @@
+    pub fn new(sponge: ArithmeticSponge<BaseField, SC>, network_id: NetworkId) -> Schnorr<SC> {
+        Schnorr::<SC> { sponge, network_id }
+    }
+
+    fn domain_bytes<S>(network_id: NetworkId) -> Vec<u8>
+    where
+        S: Signable,
+    {
+        let mut domain_string = S::domain_string(network_id);
+        // Domain prefixes have a max length of 20 and are padded with '*'
+        assert!(domain_string.len() <= 20);
+        domain_string = &domain_string[..std::cmp::min(domain_string.len(), 20)];
+        let mut bytes = format!("{:*<20}", domain_string).as_bytes().to_vec();
+        bytes.resize(32, 0);
+
+        bytes
+    }
+
+    // This function uses a cryptographic hash function to create a uniformly and
+    // randomly distributed nonce.  It is crucial for security that no two different
+    // messages share the same nonce.
+    fn blinding_hash<H>(&self, kp: &Keypair, input: H) -> ScalarField
+    where
+        H: Hashable,
+    {
+        let mut hasher = VarBlake2b::new(32).unwrap();
+
+        let mut roi: ROInput = input.to_roinput();
+        roi.append_field(kp.public.to_point().x);
+        roi.append_field(kp.public.to_point().y);
+        roi.append_scalar(kp.secret.to_scalar());
+        roi.append_bytes(&[self.network_id.into()]);
+
+        hasher.update(roi.to_bytes());
+
+        let mut bytes = [0; 32];
+        hasher.finalize_variable(|out| bytes.copy_from_slice(out));
+        // Drop the top two bits to convert into a scalar field element
+        //   N.B. Since the order p is very close to 2^m for some m, truncating only creates
+        //   a tiny amount of bias that should be insignificant and keeps the implementation
+        //   simple by avoiding reduction modulo p.
+        bytes[bytes.len() - 1] &= 0b0011_1111;
+
+        ScalarField::from_random_bytes(&bytes[..]).expect("failed to create scalar from bytes")
+    }
+
+    // This function uses a cryptographic hash function (based on a sponge construction) to
+    // convert the message to be signed (and some other information) into a uniformly and
+    // randomly distributed scalar field element.  It uses Mina's variant of the Poseidon
+    // SNARK-friendly cryptographic hash function.
+    // Details: <https://github.com/o1-labs/cryptography-rfcs/blob/httpsnapps-notary-signatures/mina/001-poseidon-sponge.md>
+    fn message_hash<S>(&mut self, pub_key: &PubKey, rx: BaseField, input: S) -> ScalarField
+    where
+        S: Signable,
+    {
+        let mut roi: ROInput = input.to_roinput();
+        roi.append_field(pub_key.to_point().x);
+        roi.append_field(pub_key.to_point().y);
+        roi.append_field(rx);
+
+        // Set sponge initial state (explicitly init state so signer context can be reused)
+        // N.B. Mina sets the sponge's initial state by hashing the input's domain bytes
+        self.sponge.state = vec![BaseField::zero(); self.sponge.state.len()];
```

Comment:
can you implement a `reset` function on `ArithmeticSponge` instead?

Reply:
Agreed - can this be part of this PR?